### PR TITLE
Automate uploading artifacts to GitHub releases

### PR DESCRIPTION
**Pull request by timja**
See [Weekly test](https://github.com/timja/jenkins/releases/tag/jenkins-2.273)
See [LTS test](https://github.com/timja/jenkins/releases/tag/jenkins-2.272)
See [Missing artifact](https://github.com/timja/jenkins/releases/tag/jenkins-2.274)

This is published when someone publishes the GitHub release notes (currently manual and done after packaging), if any artifact is missing it will continue to the next one

### Proposed changelog entries

* Entry 1: N/A
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by MarkEWaite — APPROVED** [review]
This looks great to me, especially based on the samples listed.

**Review by oleg-nenashev — APPROVED** [review]
Thanks @timja for doing it! I wanted to do the same for long time. It can help to reduce the load on our Artifactory and download mirrors while providing, and to provide an alternative download way when our download infra is down (hopefully it does not happen).

P.S: It will also require us to finish the changelog / GitHub release automation to make the flows fully automated.

**Review by res0nance — APPROVED** [review]
Love it, excited to see this on potentially the next weekly

**Comment by daniel-beck** [comment]
IIUC, GitHub has a "latest" URL that points to the latest release?

What's the plan for suspending distribution of certain releases, possibly restoring it afterwards?

**Comment by timja** [comment]
> IIUC, GitHub has a "latest" URL that points to the latest release?

It does yes

> What's the plan for suspending distribution of certain releases, possibly restoring it afterwards?

Anyone with write access can edit the release and remove uploaded files.

**Comment by timja** [comment]
If you meant how would we move the latest if we wanted to then we either delete the release or release a new one / both

File: .github/workflows/publish-release-artifact.yml
Comment (olblak) [inline .github/workflows/publish-release-artifact.yml:3]:
```suggestion
on:
  workflow_dispatch:
```

Comment (olblak) [reply .github/workflows/publish-release-artifact.yml:3]: If assets are not available from get.jenkins.io when the github action is trigger, this would allow us to force build

**Comment by olblak** [comment]
We must be clear upfront what `latest` means here. Will it be weekly or lts? knowning that latest is a symlink to the latest release so probably Weekly releases. Anyway I think 'latest' release may be confusing for end users.

A suggestion would be to use `release` and `pre-release` to map respectively lts and weekly releases? as a first iteration, I would just consider weekly releases here.
Also do we consider publishing rc release here as well?

I don't know if we can just remove the "latest" release.

Overall I like very much the approach to fetching assets from `get.jenkins.io` instead of rebuilding those here.
So it remains independant of the build process and it offers a fallback to get.jenkins.io. 
Even though debian/redhat/suse users will more than likely rely on pkg.jenkins.io which depends on get.jenkins.io

At least github releases provide download stats, so we can identify if there is an interest. Keep in mind that github release stats doesn't provide information per asset.

File: .github/workflows/publish-release-artifact.yml
Comment (olblak) [reply .github/workflows/publish-release-artifact.yml:3]: IMHO the most important thing is to be able to force a job

**Review by olblak — APPROVED** [review]
While I think we need a way to force a rebuild, overall this PR is very awesome.

**Comment by daniel-beck** [comment]
> If you meant how would we move the latest if we wanted to then we either delete the release or release a new one / both

We can (and have, several times) suspend distribution of particularly broken releases via update sites. I am asking what the corresponding process with this change is.

**Comment by olblak** [comment]
> > If you meant how would we move the latest if we wanted to then we either delete the release or release a new one / both
> 
> We can ([and have, several times](https://github.com/jenkins-infra/update-center2/blob/dd562f8ebbf2b62b86de6d8f8192d6b6fa1dbd1b/resources/artifact-ignores.properties#L307-L311)) suspend distribution of particularly broken releases via update sites. I am asking what the corresponding process with this new distribution method is.

Someone with the right permission can just delete assets and reimport them later if needed

**Comment by daniel-beck** [comment]
> Someone with the right permission can just delete assets and reimport them later if needed

Will that break automation using the 'latest' URL because there's no longer a artifact attached?

**Comment by olblak** [comment]
Yes it will,we could also probably flag a release as pre-release to remove it from 'latest', again latest won't allow you to identify latest weekly and latest stable

**Comment by timja** [comment]
> While I think we need a way to force a rebuild, overall this PR is very awesome.

Anyone with write access can re-trigger it:

image

**Comment by timja** [comment]
> > If you meant how would we move the latest if we wanted to then we either delete the release or release a new one / both
> 
> We can ([and have, several times](https://github.com/jenkins-infra/update-center2/blob/dd562f8ebbf2b62b86de6d8f8192d6b6fa1dbd1b/resources/artifact-ignores.properties#L307-L311)) suspend distribution of particularly broken releases via update sites. I am asking what the corresponding process with this new distribution method is.

The cleanest way is to mark it as a pre-release as olblak said :+1:

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: .github/workflows/publish-release-artifact.yml
Comment (jglick) [inline .github/workflows/publish-release-artifact.yml:35]: Recommend moving this sort of thing to separate `*.sh` files for legibility. (You have already run `actions/checkout`.)

File: .github/workflows/publish-release-artifact.yml
Comment (jglick) [inline .github/workflows/publish-release-artifact.yml:57]:
```suggestion
          wget -O $FILE_NAME -q https://get.jenkins.io/${REPO}/${PROJECT_VERSION}/jenkins.war
```

for robustness/clarity?

Comment (timja) [reply .github/workflows/publish-release-artifact.yml:57]: wget will always use the file name, not sure it makes much different

Comment (jglick) [reply .github/workflows/publish-release-artifact.yml:57]: Alternately,
```suggestion
          wget -q https://get.jenkins.io/${REPO}/${PROJECT_VERSION}/${FILE_NAME}
```

File: .github/workflows/publish-release-artifact.yml
Comment (timja) [reply .github/workflows/publish-release-artifact.yml:35]: this blocking?

Comment (jglick) [reply .github/workflows/publish-release-artifact.yml:35]: No, just a suggestion.

**Comment by timja** [comment]
It ran:
https://github.com/jenkinsci/jenkins/runs/1480325317?check_suite_focus=true

image

Typo fix for opensuse: https://github.com/jenkinsci/jenkins/pull/5093

**Pull request merged**
